libs/langchain/langchain/llms/watsonxllm.py

```diff
@@ -0,0 +1,172 @@
+        else:
+            if (
+                "token" not in values["credentials"]
+                and "TOKEN" not in os.environ
+                and "password" not in values["credentials"]
+                and "PASSWORD" not in os.environ
+                and "apikey" not in values["credentials"]
+                and "WATSONX_APIKEY" not in os.environ
+            ):
+                raise ValueError(
+                    "Did not find 'token', 'password' or 'apikey',"
+                    " please add an environment variable"
+                    " `TOKEN`, 'PASSWORD' or 'WATSONX_APIKEY' which contains it,"
+                    " or pass 'token', 'password' or 'apikey'"
+                    " as a named parameter in `credentials`."
+                )
+            elif "token" in values["credentials"] or "TOKEN" in os.environ:
+                values["credentials"]["token"] = get_from_dict_or_env(
+                    values["credentials"], "token", "TOKEN"
+                )
+            elif "password" in values["credentials"] or "PASSWORD" in os.environ:
+                values["credentials"]["password"] = get_from_dict_or_env(
+                    values["credentials"], "password", "PASSWORD"
+                )
+                values["credentials"]["username"] = get_from_dict_or_env(
+                    values["credentials"], "username", "USERNAME"
+                )
+            elif "apikey" in values["credentials"] or "WATSONX_APIKEY" in os.environ:
+                values["credentials"]["apikey"] = get_from_dict_or_env(
+                    values["credentials"], "apikey", "WATSONX_APIKEY"
+                )
+                values["credentials"]["username"] = get_from_dict_or_env(
+                    values["credentials"], "username", "USERNAME"
+                )
+
+        try:
+            from ibm_watson_machine_learning.foundation_models import Model
+
+            watsonx_model = Model(
+                model_id=values["model_id"],
+                credentials=values["credentials"],
+                params=values["params"],
+                project_id=values["project_id"],
+                space_id=values["space_id"],
+                verify=values["verify"],
+            )
+            values["watsonx_model"] = watsonx_model
+
+        except ImportError:
+            raise ImportError(
+                "Could not import ibm_watson_machine_learning python package. "
+                "Please install it with `pip install ibm_watson_machine_learning`."
+            )
+        return values
+
+    @property
+    def _identifying_params(self) -> Mapping[str, Any]:
+        """Get the identifying parameters."""
+        return {
+            **{"model_id": self.model_id},
+            **{"params": self.params},
+            **{"project_id": self.project_id},
+            **{"space_id": self.space_id},
+        }
+
+    @property
+    def _llm_type(self) -> str:
+        """Return type of llm."""
+        return self.llm_type
```

Comment:
this is meant to be a constant for a given class, shouldn't be user configurable

Reply:
Done, added constant 'IBM watsonx.ai'